### ISSUE TITLE
$by_correlation_id projection faulted after upgrade to Event Store 5.0.1 

### ISSUE THREAD

**Issue by Narvalex** [EventStore/EventStore#1938]
Description:
The '$by_correlation_id' projection faults after upgrading from v5.0.0 to v5.0.1, even after following the upgrade instruction found in the official blog (we stopped the projections in version 5.0.0 before the rolling upgrade).

1. Follow the upgrade instructions from version 5.0.0 to 5.0.1 found in the blog.
2. After some time (it happened al least 3 times to us) the projection will fail, like when not following the update instructions.

Expected behavior
The projection should not fail, even after reseting the projection $by_correlation_id

Actual behavior
The projection fails, and we can not recover the $bc-corrId streams with, for instance, reseting the projection.

Screenshot:
![image](https://user-images.githubusercontent.com/11648387/58730567-8761cc80-83ba-11e9-8241-d13a4b1d2da5.png)

EventStore details:
 - EventStore server version: 5.0.1

 - Operating system:
Windows 10 pro, v 1809

Additional context
The database is small, like 11.000 events.

**Comment by Narvalex** [comment]
We are willing to provide the full database to reproduce the error, as this happens consistently every time we try to upgrade from v5.0.0 to v5.0.1

**Comment by Narvalex** [comment]
We have just tried to upgrade **without** following the instructions (stop the projections before upgrading) and now works, aparently... wich is wierd.

**Comment by Narvalex** [comment]
> We have just tried to upgrade **without** following the instructions (stop the projections before upgrading) and now works, aparently... wich is wierd.

Update: It did not work every time. We are rolling back to version 5.0.0 until this issue is fixed.

**Comment by Narvalex** [comment]
I was watching carefully the stream $projections-$by_correlation_id-checkpoint, and **only** after watching it was updated I shutdown the server and proceed to do the rolling upgrade on our single node DB. And It worked! I will close this issue for now, but I will reopen if any issues arise again. Thanks @shaan1337 for your careful observation

**Comment by Narvalex** [comment]
It did happen again. Bellow the screenshot
![image](https://user-images.githubusercontent.com/11648387/58840520-ebdb9080-8633-11e9-8fbd-9907837537b2.png)

**Comment by Narvalex** [comment]
> It did happen again. Bellow the screenshot
This time, I did followed the instructions. I appeared to work. But as soon as I restarted the node, it happened again 😟

**Comment by shaan1337** [comment]
@Narvalex hello, thanks for all the details, we're trying to reproduce this issue.

Would you please be able to upload your database somewhere and send me a link to shaan.nobee [ a t ] eventstore.org? (S3 or Google drive should be OK) This will definitely help us fix this bug more quickly.

Thanks!

**Comment by Narvalex** [comment]
Mail sent! Thanks @shaan1337 !

**Comment by shaan1337** [comment]
As discussed with @Narvalex offline, it appears that setting the eventstore config option `StartStandardProjections` to `true` may be causing the above issues.

**Comment by Narvalex** [comment]
Thanks @shaan1337 . I will reopen if it happens again

**Comment by Narvalex** [comment]
I did happen again. Now the projection is faulted in production, as I already mentioned to @shaan1337 .

**Comment by Narvalex** [comment]
The pull request #1882 is the cause for this issue, at least with following projection:
``` js
fromAll()
    .when({ 
        $any: (s, e) => {
            if (e.streamId.substring(0, 1) === '$')
                return;

            linkTo('all', e);
       }
})
```

**Comment by Narvalex** [comment]
It happens after the node is restarted

**Comment by YoEight** [comment]
Should be fixed by #2447

**Comment by Narvalex** [comment]
I will test ASAP and let you know, @YoEight

**Comment by Narvalex** [comment]
Thanks again @YoEight for your patience. The #2447 does prevent the `$by_correlation_id` to get faulted, indeed, but, given the scenario described in #1952, it does not appear any new written events with corrIds in the `$bc-[correlationId]` streams. 
Only after merging with #1952 the new events appears again in the `$bc-[corId]` streams.

**Comment by YoEight** [comment]
@Narvalex The patch I did was about discarding a link event targetting an event that no longer exists. It can happen if a stream has a max age or max count set in its metadata and that targetted event got scavenged. Is this description fits your situation?

**Comment by Narvalex** [comment]
@YoEight No, it does not fit my situation. My situation is about a projection like the following

> 
> ```js
> fromAll()
>     .when({ 
>         $any: (s, e) => {
>             if (e.streamId.substring(0, 1) === '$')
>                 return;
> 
>             linkTo('all', e);
>        }
> })
> ```

When the aformetioned projection runs, and after a database restart the $bc-[correlationId] streams does not make progress anymore. Your path indeed prevents the $by_correlation_id to become faulted, buth still the $bc-[correlationId] does not make progress anymore. 
You can reproduce this error by following the description of pull request #1952

**Comment by YoEight** [comment]
I asked you that question specifically because if my patch prevents `$by_correlation_id` from becoming faulted, it means that event wasn't supposed to be written. About `$by_correlation_id`, are you sure it doesn't make any more progress? Let me be clearer, are you certain `$by_correlation_id` has events to process? If you take that information from the UI, it is possible you are reading a cached value if there are no incoming events for `$by_correlation_id` to process. The UI is eventually consistent when reading that information from the database.

**Comment by Narvalex** [comment]
I am certain about that. You can verify for yourself. Is pretty easy. I recently did for double checking. 

1. Since your patch is merged with master I pulled master and builded
2. Then I started the databse (is empty)
3. Created a continous projection of the above projection with emit and track enabled.
4. Commited events with metadata: 

```
{
    "correlationId": "123"
}
```

5. The `$by_correlation_id` is not faulted, but NO events are on the $bc-123 stream. The stream itself was not created.
6. I am aware of the eventual consistency nature of the UI, but I disabled the cache on my browser, waited a lot, and to prove it...
7. I merged locally with #1952
8. And the $bc-123 appears with all its events

**Comment by Narvalex** [comment]
I'am certain. Double checked with fresh build from master (since your patch is already merged with it). With cache disabled in my browser I can see no progress in `$bc-[correlationId]`. After merging with #1952 the `$by_correlation_id` makes progress.

**Comment by Narvalex** [comment]
Hi @YoEight, thanks for the pull request #2465. Unfortunately, I have to inform you that the patch does not solve the problem of this issue. Also, I have found that it invalidates patch #1952, which means that the `$by_correlation_id` projection makes no further progress. 

I understand that it is best to solve the root problem, but, in any case, in our organization we depend on the continuous projection that filters the events with prefix "$" and the system projection `$by_correlation_id`, which does not work without the Fix #1952.
